src/control/tools/SplineHandler.cpp

```diff
@@ -77,123 +94,115 @@ auto SplineHandler::onKeyEvent(GdkEventKey* event) -> bool {
-    double h = stroke->getElementHeight() + 2 * radius;
+    Rectangle rect = this->computeRepaintRectangle();
 
     switch (event->keyval) {
         case GDK_KEY_Escape: {
-            if (pointCount > 0) {
-                // remove dynamically changing point at cursor position
-                stroke->deletePoint(pointCount - 1);
-                this->redrawable->rerenderRect(x, y, w, h);
-            }
-            finalizeSpline();
+            this->redrawable->repaintRect(rect.x, rect.y, rect.width, rect.height);
+            this->finalizeSpline();
             return true;
         }
         case GDK_KEY_BackSpace: {
-            if (pointCount > 1) {
-                // delete last non dynammically changing point
-                stroke->deletePoint(pointCount - 2);
+            if (this->getKnotCount() > 0) {
+                this->deleteLastKnotWithTangent();
             }
             break;
         }
-        case GDK_KEY_Up: {
-            movePoint(0, -1);
+        case GDK_KEY_x:
+        case GDK_KEY_X: {  // x like "x-direction"
+            double xshift = (event->state & GDK_SHIFT_MASK) ? -1 : 1;
+            this->movePoint(xshift, 0);
             break;
         }
-        case GDK_KEY_Down: {
-            movePoint(0, 1);
+        case GDK_KEY_y:
+        case GDK_KEY_Y: {  // y like "y-direction"
+            double yshift = (event->state & GDK_SHIFT_MASK) ? 1 : -1;
+            this->movePoint(0, yshift);
             break;
         }
-        case GDK_KEY_Right: {
-            movePoint(1, 0);
+        case GDK_KEY_r:
+        case GDK_KEY_R: {  // r like "rotate"
+            double angle = (event->state & GDK_SHIFT_MASK) ? -5 : 5;
+            double xOld = tangents.back().x;
+            double yOld = tangents.back().y;
+            double xNew = cos(angle * M_PI / 180) * xOld + sin(angle * M_PI / 180) * yOld;
+            double yNew = -sin(angle * M_PI / 180) * xOld + cos(angle * M_PI / 180) * yOld;
+            this->modifyLastTangent(Point(xNew, yNew));
             break;
         }
-        case GDK_KEY_Left: {
-            movePoint(-1, 0);
+        case GDK_KEY_s:
+        case GDK_KEY_S: {  // s like "scale"
+            double xOld = tangents.back().x;
+            double yOld = tangents.back().y;
+            double length = 2 * sqrt(pow(xOld, 2) + pow(yOld, 2));
+            double factor = 1;
+            if ((event->state & GDK_SHIFT_MASK) && length >= 1) {
+                factor = 1 / 1.1;
+            } else if (!(event->state & GDK_SHIFT_MASK) && length <= 2000) {
+                factor = 1.1;
+            }
+            double xNew = xOld * factor;
+            double yNew = yOld * factor;
+            this->modifyLastTangent(Point(xNew, yNew));
             break;
         }
     }
-    this->redrawable->rerenderRect(x - 1, y - 1, w + 2, h + 2);
+    this->redrawable->repaintRect(rect.x - 1, rect.y - 1, rect.width + 2, rect.height + 2);
     return false;
```

Comment:
This always returns false, which means that the event will continue to be propagated even when it has been handled here. You will need to return true if you've done something with the keypress.

Reply:
I switched back to arrow keys and corrected the return statements.